### ISSUE TITLE
Sort components and tools in crafting recipe QoL

### ISSUE THREAD

**Issue by Brambor** [CleverRaven/Cataclysm-DDA#25215]
In the following text, I write only crafting components, but the same is true for required tools.

**Problem description.**  
Often json files have a different order of components (relative one to another), in game the order is the same as in json. However, when crafting, it would be more convenient, if they would always be in the same order (relative one to another), that might be alphabetical.
The reason for that is comparing recipes, one example, are these two:
```
  {
    "type": "recipe",
    "result": "rifle_22",
    "category": "CC_WEAPON",
    "subcategory": "CSC_WEAPON_RANGED",
    "skill_used": "mechanics",
    "skills_required": [ "gun", 1 ],
    "difficulty": 3,
    "time": 30000,
    "reversible": true,
    "autolearn": true,
    "qualities": [ { "id": "SAW_M_FINE", "level": 1 }, { "id": "SCREW_FINE", "level": 1 } ],
    "components": [ [ [ "pipe", 1 ] ], [ [ "2x4", 1 ] ], [ [ "scrap", 2 ] ] ]
  },{
    "type": "recipe",
    "result": "pipe_double_shotgun",
    "category": "CC_WEAPON",
    "subcategory": "CSC_WEAPON_RANGED",
    "skill_used": "mechanics",
    "skills_required": [ "gun", 1 ],
    "difficulty": 3,
    "time": 18000,
    "reversible": true,
    "autolearn": true,
    "qualities": [ { "id": "SAW_M_FINE", "level": 1 }, { "id": "SCREW_FINE", "level": 1 } ],
    "components": [ [ [ "pipe", 2 ] ], [ [ "scrap", 3 ] ], [ [ "2x4", 1 ] ] ]
  },
```
in the first one, the order is `"components": [ [ [ "pipe", 1 ] ], [ [ "2x4", 1 ] ], [ [ "scrap", 2 ] ] ]` while in the other it is `"components": [ [ [ "pipe", 2 ] ], [ [ "scrap", 3 ] ], [ [ "2x4", 1 ] ] ]`

**The solution I'd like**
Sort components on the fly.
This would mean that recipes from new contributions to json would be automatically sorted, as well as recipes in all mods etc.
My question is, is this a performance a problem? I believe it shouldn't, if only one recipe description is loaded at time, then only one needs sorting. That way, the number of elements is always such a small pool.

**Alternatives I've considered** 
The other way I considered is to sort it in json files.
I started to make a python program for that, but it might be better, to just do it on the fly (^^^). This would require editing nearly all recipis (could be automated) and then needs to be reviewed.

**Additional context**  
The program is not ready yet, doesn't do much, but I can provide it once some more work has been done. But I would like suggested solution to be implemented.

I would like to work on that. I have never done anything in C and never compiled anything, but the issue might be solved by a change of one line, so I might just manage that. My question is where would I look for the code handling displaying crafting recipes.

**Comment by alanbrady** [comment]
Doing some kind of sort on the components as their imported during runtime seems reasonable to me.

**Comment by CoroNaut** [comment]
Just remember that sorting the lists during runtime is extra work for the computer, simply sorting them out in json cuts out this extra work entirely. I am for that reason against just calculating the order when you select a recipe.

**Comment by alanbrady** [comment]
> Just remember that sorting the lists during runtime is extra work for the computer, simply sorting them out in json cuts out this extra work entirely. I am for that reason against just calculating the order when you select a recipe.

You're talking about optimizing code that is not in the hot path.

> And with this, creating a script to sort items would allow any new item to be put into the game with just one extra step.

The problem here is you can't do this automated as a part of the PR process.  You can make a script to check if a commit adds new stuff that is out of order and fail the build, but you can't easily and correctly retroactively apply a commit after a PR to fix someone's file.  That's why the linter emits build failures when you submit something with bad formatting instead of linting it itself.

Then you also have to talk about all the thrash that forcing a certain format creates.  There's a reason we haven't just astyled the whole codebase in one go.

**Comment by CoroNaut** [comment]
I am fine with you doing it on runtime for exactly the reasons you described. I am just against it as it is work being done that really does not need to be done. 

Example in a different scenario: Imagine just going out to buy groceries every time you have a meal. You waste all the gas going back and forth when you could just get lots of groceries after eating lots of meals all at the same time. Its perfectly acceptable both ways, but doing it all at once is much more efficient.

**Comment by kevingranade** [comment]
It's a trivial amount of work to sort these short lists in the game  but it's not a trivial amount of work for maintainers to deal with thousands of lines of json shuffling around.

**Comment by CoroNaut** [comment]
It is also trivial to make a script to alphabetize a list and to find these lists using regex making the whole process a couple steps, then running the script and it can just alphabetize all recipes in all files.

**Comment by ZhilkinSerg** [comment]
Reopening as it was not fully fixed by #25562 yet.

**Comment by Brambor** [comment]
Alternatively it could be sorted by id which is the same for all languages.
I saw somewhere that it could be problem to sort for Chinese, so this would be consistent.
Just throwing it out there so it is known.
This issue still requests alphabetical order thougth.

**Comment by BevapDin** [comment]
I thought about sorting that, but I did not see a good sorting algorithm:
What if recipe A needs "1 leather patch or 1 rag" and "1 heavy stick or 1 two-by-four", while another needs "1 rag or 1 heavy stick" and "1 ammo belt".

The first would get this list:
```
    1 heavy stick or 1 two-by-four
and 6 leather patches or 1 tanned hide
```
And the second would get 
```
    1 tanned hide or 1 two-by-four
and 1 wired sheet
```
The items used in both recipes are at different positions in the list, it does not seem very useful.


Or this:
```
1 heavy stick or 1 long stick or 1 two-by-four
... (more components)
```
vs
```
... (more components)
1 two-by-four
```
The two-by-four (starting with a `t`) would be listed near the end, while in the first recipe it would be listed near the start (line starts with `h` for heavy stick).

In short: items will appear at random places in the list, depending on the other required items. I don't see this automatic sorting as useful.

----

@Brambor we don't sort by untranslated strings, see https://github.com/CleverRaven/Cataclysm-DDA/pull/21257#issuecomment-314960543

**Comment by ZhilkinSerg** [comment]
@BevapDin That totally makes sense. Thanks.

**Comment by Brambor** [comment]
@BevapDin Well it is more likely to have items of the same kind to be used instead of (e.g. `two-by-four OR heavy stick` and `rag OR leather patch`).
This can be easily seen by understanding that the purpose of the OR in the crafting recipe is that the items share some quality (`is a stick` in components, `is a hammer` in both components and tools).
And therefore there are lots of recipies that use any of the sticks as handle, any of the plastic containers to make bigger container, any of the rags to make sheet etc.

However there might be cases that you are indicating. I believe that this is small portion of the recepies. These would not benefit from alphabetical sorting, but it wouldn't hurt them either.
I'm not sure, if there are any (thougth it is probable) and even if there are none, they might be added in the future, so we have to think about them anyway.

So all in all.
- **(1)** It makes the components of majority (I guess) recipes more consistent and that helps.
- **(2)** It might make some, if any, recipies look unsorted or ordered in unintuitive order. This is the opposite of (1), but as there are more cases in (1) than there are in (2), the overall effect is positive.

**If we cannot agree on the sizes of sets presented, then we could actually count or make some examples of both, maybe just a small sample.** I believe, that weapons and such recipies tend to be more in (1) and some chemical recipies could be in (2). There is one example in issue description.

**Comment by Brambor** [comment]
However there is the issue of splitting `apple OR irradiated apple OR blueberry or irradiated blueberry`. They would still have the alphabetical order thougth.
